lambda/entry.go

```diff
@@ -18,31 +18,73 @@ import (
 //   - handler may take between 0 and two arguments.
 //   - if there are two arguments, the first argument must satisfy the "context.Context" interface.
 //   - handler may return between 0 and two arguments.
-//   - if there are two return values, the second argument must be an error.
+//   - if there are two return values, the second return value must be an error.
 //   - if there is one return value it must be an error.
 //
 // Valid function signatures:
 //
 //	func ()
+//	func (TIn)
 //	func () error
 //	func (TIn) error
 //	func () (TOut, error)
 //	func (TIn) (TOut, error)
-//	func (context.Context) error
+//	func (context.Context, TIn)
 //	func (context.Context, TIn) error
-//	func (context.Context) (TOut, error)
 //	func (context.Context, TIn) (TOut, error)
 //
 // Where "TIn" and "TOut" are types compatible with the "encoding/json" standard library.
 // See https://golang.org/pkg/encoding/json/#Unmarshal for how deserialization behaves
-func Start(handler interface{}) {
+func Start(handler any) {
 	StartWithOptions(handler)
 }
 
+// StartWithOptionsTypeSafe is the same as StartWithOptions except that it takes a generic input
+// so that the function signature can be validated at compile time.
+// The caller can supply "any" for TIn or TOut if the input function does not use that argument or return value.
+//
+// Examples:
+//
+// TIn and TOut ignored
+//
+//	StartWithOptionsTypeSafe[any, any](func() {
+//		fmt.Println("Hello world")
+//	})
+//
+// TIn used and TOut ignored
+//
+//	type event events.APIGatewayV2HTTPRequest
+//	StartWithOptionsTypeSafe[event, any](func(e event) {
+//		fmt.Printf("Version: %s", e.Version)
+//	})
+//
+// TIn ignored, TOut used and an error returned
+//
+//	type resp events.APIGatewayV2HTTPResponse
+//	StartWithOptionsTypeSafe[any, resp](func() (resp, error) {
+//		return resp{Body: "hello, world"}, nil
+//	})
+func StartWithOptionsTypeSafe[TIn any, TOut any, H HandlerFunc[TIn, TOut]](handler H, options ...Option) {
+	start(newHandler(handler, options...))
+}
+
+// HandlerFunc represents a valid input as described by Start
+type HandlerFunc[TIn, TOut any] interface {
+	func() |
+		func(TIn) |
+		func() error |
+		func(TIn) error |
+		func() (TOut, error) |
+		func(TIn) (TOut, error) |
+		func(context.Context, TIn) |
+		func(context.Context, TIn) error |
+		func(context.Context, TIn) (TOut, error)
```

Comment:
Yeah I agree it'll be much nicer to help go infer the types and avoid passing in unused types.

Makes the naming problem a bit harder though 😄

Reply:
Hmm so the type inference doesn't seem to want to play ball https://go.dev/play/p/Ls5MrkTK76g

I could split up the interface so at least the caller doesn't need to specify a type that they don't use. What do you reckon?

Reply:
I'd be OK with 
```
type HandlerFunc interface {
    func(context.Context, TIn) (TOut, error)
}
```
since then the inference on TIn/TOut works https://go.dev/play/p/r3TIVrSuc4j

I'm not super attached to the polymorphic-ish style that `Start`'s reflection code currently supports, if Go had generics in 2017 I'd probably not have done it that way

> I could split up the interface so at least the caller doesn't need to specify a type that they don't use

I'm not sure what you mean by this, got an example?

Reply:
Type inference seems to only work if the interface only has a single variant.

In this example the user needs to supply a type for `I`, despite not using it https://go.dev/play/p/-7nPycevowb

I was suggesting having more interfaces, so that could be avoided:

```
// more types more things to name though (:
type HandlerFunc interface {
	func() | func() error
}

type HandlerFuncTIn[I any] interface {
	func(I) | func(I) error | func(context.Context, I) | func(context.Context, I) error
}

type HandlerFuncTOut[O any] interface {
	func() (O, error)
}

type HandlerFuncTInTOut[I, O any] interface {
	func(context.Context, I) (O, error)
}
```

But then that's four interfaces and doesn't solve the inference problem (except for HandlerFuncTOut). So maybe one interface per variant...

Reply:
Thanks for explaining! And yeah, things like `func(I) | func(I) error | func(context.Context, I) | func(context.Context, I) error` from your example _looks like_ it should work, but still seem to result in inference errors in the playground - maybe we're just too early to the generics party 🤷‍♂️

Reply:
A human can certainly infer them! So yeah maybe it'll work in the future.

For now, I'd be happy to at least have some programmatic way of checking the function is valid. Perhaps exporting validateArguments and validateArguments or a new function that encapsulated them both?

Reply:
What's the use case you have for exporting the validation logic?

Reply:
So the original goal of this PR was to have assurance that the function supplied to lambda.Start is valid without having to compile and run and invoke the running lambda. It's shown in this PR this can be achieved at compile time with generics, but I'm guessing that the inference issue means that you're probably not keen to have that monolithic interface or the 4+ smaller ones.

Exporting the validation logic gives an easy path to validate the function in a unit test, or I suppose by checking and panicking early at runtime. Plus it's a much smaller change.

What do you reckon?

Reply:
>  but I'm guessing that the inference issue means that you're probably not keen to have that monolithic interface or the 4+ smaller ones.

I am still good with a single interface (ref: https://github.com/aws/aws-lambda-go/pull/468#discussion_r1014384177) - if the inference gets smarter in the future, then it could be grown to look like the your original monolithic interface

> Exporting the validation logic gives an easy path to validate the function in a unit test, or I suppose by checking and panicking early at runtime.

:) I think for this you'd want to rely on the unit test (or the generic type, when added!) rather than `panic`-ing at runtime. A panic triggers a sort of cold start, which under load could contribute to consuming the account's Lambda concurrency limit.

Reply:
> I am still good with a single interface (ref: https://github.com/aws/aws-lambda-go/pull/468#discussion_r1014384177) - if the inference gets smarter in the future, then it could be grown to look like the your original monolithic interface

So just to confirm, you're saying you're happy adding just that single interface with the single variant? So the plan would be to start with adding that to the public interface, and then when the inference gets better, adding support for the other variants?
```go
type HandlerFunc interface {
    func(context.Context, TIn) (TOut, error)
}
```

> :) I think for this you'd want to rely on the unit test (or the generic type, when added!) rather than panic-ing at runtime. A panic triggers a sort of cold start, which under load could contribute to consuming the account's Lambda concurrency limit.

100% agree. I'm never a fan of panics. session.Must never seems right to me

Reply:
> So just to confirm, you're saying you're happy adding just that single interface with the single variant? So the plan would be to start with adding that to the public interface, and then when the inference gets better, adding support for the other variants?

Yup!